[postgres] Add connection setting for restricting schema

nyalldawson opened the pull request:
Like the equivalent Oracle provider setting, this allows users to specify a single schema to limit a Postgres connection to.

When set, only tables from the matching schema will be included in the browser panel and data source select for the connection.

This can be used to limit the database work required to populate tables for a connection pointing to a large database store.

Sponsored by Regiodata

![image](https://github.com/user-attachments/assets/375aa972-c416-4fef-aaed-c6c8712a40f6)

nyalldawson committed:
Add some consistency to db new connection dialogs

- Split oracle and postgres dialogs into two columns with the
  database details in a separate group to connection details,
  to match the SQL Server connection dialog
- Fix tab ordering
- Consistently use "Test Connection" text for buttons
- Fix bad (narrow) stretch for settings in oracle connection dialog

nyalldawson committed:
[postgres] Add connection setting for restricting schema

Like the equivalent Oracle provider setting, this allows users
to specify a single schema to limit a Postgres connection to.

When set, only tables from the matching schema will be included
in the browser panel and data source select for the connection.

This can be used to limit the database work required to populate
tables for a connection pointing to a large database store.

Sponsored by Regiodata

troopa81 commented on src/providers/postgres/qgspostgresconn.cpp:
```diff
@@ -2800,6 +2814,12 @@ bool QgsPostgresConn::allowRasterOverviewTables( const QString &connName )
   return settings.value( "/PostgreSQL/connections/" + connName + "/allowRasterOverviewTables", true ).toBool();
 }
 
+QString QgsPostgresConn::restrictToSchema( const QString &connName )
```
```suggestion
QString QgsPostgresConn::schemaToRestrict( const QString &connName )
```

restrictToSchema name let's think that it actually modify something where it actually only return a schema name

nyalldawson replied:
Done

troopa81 commented on src/providers/postgres/qgspostgresconn.cpp:
```diff
@@ -1137,12 +1137,24 @@ bool QgsPostgresConn::supportedLayersPrivate( QVector<QgsPostgresLayerProperty>
 }
 
-bool QgsPostgresConn::getSchemas( QList<QgsPostgresSchemaProperty> &schemas )
+bool QgsPostgresConn::getSchemas( QList<QgsPostgresSchemaProperty> &schemas, const QStringList &restrictToSchemas )
```
this function allows to take several schemas but the UI allows only one, should we not changing this function or the UI to be both consistent ?

jgrocha replied:
Just two cents regarding the UI.
1) We could have a combobox to choose one schema from the database instead of a text field.
2) We could have a button on the UI for "Schema selection". It opens a model dialog where the user can choose which schemas to consider, like this one in DBeaver:
![imagem](https://github.com/user-attachments/assets/17348113-1e78-4c20-b716-de2055b56f27)
We could show the tables in each schema to help the user decide.

This can be made in a another PR, since the underlying functionality is already provided by this one.

nyalldawson replied:
@troopa81 
> this function allows to take several schemas but the UI allows only one, should we not changing this function or the UI to be both consistent ?

At some stage it would be good to extend this to allow selection of multiple schemas in the UI, like what we support for SQL Server connections. It's out of scope for my current work (which is just limited to matching the Oracle provider functionality), but I think it doesn't hurt to have the API ready for this.

nyalldawson replied:
@jgrocha 
I agree, but that's out of scope for my current contract.

troopa81 reviewed (commented):
Shall we not merged this option with "Only look in the public schema" ? What happen if we have both public checked and this option set ?

nyalldawson committed:
Disable schema filtering when 'Only look in 'public' schema" is checked

These two options are mutually exclusive

nyalldawson commented:
@troopa81 
> Shall we not merged this option with "Only look in the public schema" ? What happen if we have both public checked and this option set ?

I've made them mutually exclusive -- when the "only look in public schema" checkbox is ticked, the schema filter box is now disabled.

nyalldawson committed:
Rename method for clarity

nyalldawson committed:
Correctly remove outdated configuration when storing db connections

nyalldawson committed:
[api] Simplify handling of 'public schema only' option

Just use the 'restrict to schema' API with the public schema
instead of having explicit bool arguments to restrict to public
schema

nyalldawson committed:
Fix some warnings

nyalldawson committed:
[postgres] Restrict to public schema should also apply to browser schemas

nyalldawson committed:
Add placeholder text to improve UI

nyalldawson commented:
@troopa81 Ok, I've reworked the handling of "public schema only" to fit in with this, and simplified the API accordingly.

It also turns out that the "public schema only" option wasn't well handled, as it was ignored when populating the browser items for the connection. That's fixed now.
